Sanitize world names when creating them, and store the original name elsewhere

Df458 opened the pull request:
This PR addresses #4875 by preventing potentially invalid paths from being created when making a new world.

The way it works is fairly simplistic: If a character matches against a blacklist (containing path characters and chars forbidden by certain filesystems), it is replaced by '\_' in the path. The display name is stored in `world.mt`, and duplicate file names are resolved by adding an incrementing suffix (\_1, \_2, \_3, etc) as suggested in the original issue. Additionally, the world_ prefix is added if the name matches a blacklist of 'reserved names' that some filesystems have trouble with (see the original issue for more details).

## Testing
Create worlds, and put garbage into their names. Try making worlds with different names but duplicate sanitized names. Try hosting/launching these worlds. Ensure that these worlds aren't being duplicated disappearing for any reason. Check anywhere in the UI that you would expect to see the world name: Is the name you entered there, or is it the sanitized version?

## Known Issues
- Names are still case-sensitive. On case-insensitive filesystems, collisions can probably still happen (needs testing)

appgurueu commented:
> the world is not created (normal) but no feedback is given to the user about this
I would go with a plain error message saying "World could not be created: too many worlds with similar names, chose a different one."

Df458 commented:
Rebased on 5.2. I'll try to finish this feature up this week so that it can be reviewed/merged.

Df458 commented:
So I've added the necessary code to tell the user more directly when a world can't be created due to hitting the limit--but in retrospect, I'm starting to wonder if an arbitrary limit is even really necessary.

I mostly created it to eventually break out of the while loop, but aside from slowness (and difficulty manually managing the directories) there's no other real downside to creating similar worlds. We could just raise the limit to max int, which I seriously doubt anyone will manage to hit.

Any thoughts on this

paramat commented:
For me it seems better and simpler to just refuse a non-compliant world name in main menu and display a message of explanation. No need to store the display name, avoids duplicates, as well as teaching users how to properly name their worlds and so avoid repeating the problem.

Df458 commented:
> For me it seems better and simpler to just refuse a non-compliant world name in main menu and display a message of explanation.

So what's a non-compliant world name? Obviously, we should block certain characters like '/'. But on case-insensitive filesystems, I suspect we already have weirdness with lowercase/uppercase world names (which I'm pretty sure minetest doesn't check for). Do we also update our duplicate check to be case-insensitive, although some systems support them? What about special characters that aren't path separators, but are still barred from windows while being allowed on linux? Do we also explicitly check for 'NUL' while we're at it? How about on non-NTFS filesystems?

...and if the answer to all that is yes, how exactly do you propose we explain the resulting ruleset to a user? I'm not saying it's impossible, but I wouldn't call it 'simpler'.

Df458 commented:
Bump. What do other contributors and coredevs think? This PR has been open for close to a month and @paramat disapproves of the current approach, so if I don't see any support for this in the next week or so I'll close it to free up the backlog.

sfan5 commented:
Sounds okay to me in principle but Unicode is definitely a problem that has to be considered.
Perhaps it would be possible to turn just the characters that cause problems into `_` (instead of everything outisde `A-Za-z0-9_-`).

rubenwardy commented:
> ...and if the answer to all that is yes, how exactly do you propose we explain the resulting ruleset to a user? I'm not saying it's impossible, but I wouldn't call it 'simpler'.

You don't. You save worlds like `nul` as `nul1` or some other non-problematic name

Df458 commented:
> Perhaps it would be possible to turn just the characters that cause problems into _

I'm a little wary of this because I suspect we'll miss something along the way, but even an incomplete blacklist is better than no checks at all. I'll change the PR to block path-related characters and the list [here](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file?redirectedfrom=MSDN#file-and-directory-names) for starters.


 > Minetest should probably avoid duplicate world display names though, but duplicate world folder names should need to be mitigated

Pretty sure this PR still blocks duplicate display names, since there's a separate check for that. The only time where this could happen is if a user manually changes the display name in a text editor, or downloads a world with a duplicate name but non-duplicate directory name. I think those edge-cases are fine to leave, since they should be pretty uncommon.

I still think we should keep the `world_` prefix as-is, since it prevents cases like `nul` without adding a bunch of special checks into the code.

Does this sound acceptable?

rubenwardy commented:
That sounds good to me

sfan5 commented:
> I still think we should keep the `world_` prefix as-is, since it prevents cases like `nul` without adding a bunch of special checks into the code.

I often type out the filesystem paths to my worlds and would not like to be inconvenienced by Microsoft's implementation choices when I'm not even using Windows.
So my preference would be to keep an explicit blacklist of the (`con`, `nul`, ...) filenames for Windows. I don't mind if `:` turns to `_` even on Linux, just the `world_` prefix is what annoys me.

Df458 commented:
I've replaced the whitelist with a blacklist as requested. I also ran some tests and modified the function to handle Unicode characters better. I'm not well-versed in handling Unicode in C/C++, so please let me know if I've done anything monumentally stupid by accident.

> So my preference would be to keep an explicit blacklist of the (con, nul, ...) filenames for Windows.

What do you propose we do when we encounter a blacklisted name? Should we add the `world_` prefix then, or do something else?

Df458 commented:
One more thing: I'd like someone on Windows to test what happens when making two worlds with the same letters but different casing (ie 'test' and 'TEST'). Please let us know what happens, I have a hunch but I'd prefer actual data for this case.

sfan5 commented:
> What do you propose we do when we encounter a blacklisted name? Should we add the world_ prefix then, or do something else?

Considering it's better when it "just works", adding the `world_` prefix is a better approach.

Df458 commented:
@sfan5 / @rubenwardy -- I believe I've implemented what was asked of me. If you've got the time, can you take a look and confirm if things are working as you'd like them to now?

rubenwardy commented on src/script/lua_api/l_mainmenu.cpp:
```diff
@@ -618,16 +618,17 @@ int ModApiMainMenu::l_create_world(lua_State *L)
-			lua_pushstring(L, "Failed to initialize world");
+		std::string error;
+		if (!loadGameConfAndInitWorld(path, name, games[gameidx], true, &error)) {
+			lua_pushstring(L, std::string("Failed to initialize world: " + error).c_str());
```
the string construction here is redundant, you probably meant 

```c++
(std::string("Failed to initialize world: ") + error).c_str()
```

which can be simplified to

```cpp
("Failed to initialize world: "s + error).c_str()
```

But given the code in your PR compiles, this is somehow OK:

```cpp
("Failed to initialize world: " + error).c_str()
```

Df458 replied:
Fixed with the first option. The `""s` notation doesn't compile on my end for whatever reason, but you're correct that I missed those parens.

rubenwardy commented on src/server.cpp:
```diff
@@ -360,8 +360,11 @@ void Server::init()
-		throw ServerError("Failed to initialize world");
+	std::string error;
+	if (!loadGameConfAndInitWorld(m_path_world,
+				fs::GetFilenameFromPath(m_path_world.c_str()),
```
indent too far

rubenwardy commented on src/util/string.cpp:
```diff
@@ -889,3 +889,91 @@ std::wstring translate_string(const std::wstring &s)
 #endif
 }
+
+static const std::wstring disallowed_dir_names[] {
```
consider making this a `std::array` so the length becomes a compile error if it does not match

rubenwardy commented on src/util/string.cpp:
```diff
@@ -889,3 +889,91 @@ std::wstring translate_string(const std::wstring &s)
+/**
+ * List of characters that are blacklisted from created directories
+ */
+static const wchar_t disallowed_path_chars[] {
```
same here

rubenwardy commented:
If you copy and rename a world, the name stays the same in the world list as _all_ worlds are given `world_name`s. I don't necessarily think this is a problem

paramat commented:
Df458, i only had a partial understanding of the situation, so i retract my earlier suggestion.
I did not 'disapprove' of the implementation as in 'officially opposes the PR', i did not feel that strongly about it.

Df458 commented:
Rebased on 5.4.0

By the way, I don't think the "Action / change needed" label applies anymore. All the review comments were resolved a couple weeks ago.

SmallJoker commented on src/util/string.cpp:
```diff
@@ -889,3 +890,89 @@ std::wstring translate_string(const std::wstring &s)
+/**
+ * List of characters that are blacklisted from created directories
+ */
+static const std::array<wchar_t, 10> disallowed_path_chars = {
```
If preferred: This could be a simple string, without manual iteration:
```C++
static const std::wstring disallowed_path_chars = L"<>:\"/\\|?*.";

is_valid = tmp.find_first_of(disallowed_path_chars) == std::wstring::npos;
```

SmallJoker commented on src/util/string.cpp:
```diff
@@ -889,3 +890,89 @@ std::wstring translate_string(const std::wstring &s)
+ *	and add a prefix to them
+ * 2. Remove 'unsafe' characters from the name by replacing them with '_'
+ */
+std::string sanitizeDirName(const std::string &str, const std::string &unsafe_prefix)
```
`unsafe_prefix` .. the prefix is unsafe?
What would you think about `optional_prefix` or so?

SmallJoker commented on src/util/string.cpp:
```diff
@@ -889,3 +890,89 @@ std::wstring translate_string(const std::wstring &s)
+ */
+std::string sanitizeDirName(const std::string &str, const std::string &unsafe_prefix)
+{
+	std::wstring tmp = utf8_to_wide(str);
```
```suggestion
	std::wstring safe_name = utf8_to_wide(str);
```
..or something like that.

SmallJoker commented on src/util/string.cpp:
```diff
@@ -889,3 +890,89 @@ std::wstring translate_string(const std::wstring &s)
+
+	for(std::wstring disallowed_name : disallowed_dir_names) {
+		if (str_equal(tmp, disallowed_name, true)) {
+			tmp = utf8_to_wide(unsafe_prefix) + tmp;
```
`break;` below.
Missing space after `for`

SmallJoker replied:
Fixed this.

SmallJoker commented on src/server.cpp:
```diff
@@ -356,8 +356,11 @@ void Server::init()
+	std::string error;
+	if (!loadGameConfAndInitWorld(m_path_world,
+            fs::GetFilenameFromPath(m_path_world.c_str()),
+            m_gamespec, false, &error))
```
Spaces -> tabs

SmallJoker commented on src/content/subgames.cpp:
```diff
@@ -288,8 +306,27 @@ std::vector<WorldSpec> getAvailableWorlds()
+
+		if (fs::PathExists(final_path)) {
+			if (error != nullptr) {
+				*error = "Too many similar filenames";
```
Wouldn't this be an ideal use-case for exceptions?

Df458 replied:
I'm always a little wary of throwing exceptions in a gamedev context, but since this isn't a hotspot it should be fine. Do you know any existing exceptions that would fit the bill here, or would you prefer that I make a new one?

SmallJoker replied:
A generic exception (`BaseException`) likely suffices here. Yet I'll leave the decision whether to use exceptions to you.
Oh and `InvalidFilenameException` is currently unused and could be removed...

Df458 replied:
I thought this was a good suggestion at first, but after writing most of the necessary code I found that it makes the error checking in `l_mainmenu.cpp` kinda messy as a result... we could make it work by making this function return void and throwing an exception for the other failure state as well, but I'd like your opinion on that first.

SmallJoker replied:
It would be simple as
```C++
void loadGameConfAndInitWorld()
{
	throw BaseException("Some error");
}

try {
	loadGameConfAndInitWorld(path, name, games[gameidx], true))
	lua_pushnil(L);
} catch (BaseException &e) {
	lua_pushstring(L, ("Failed to initialize world: " + e.what()).c_str());
}
```

Df458 replied:
Gotcha, I'll go ahead and replace both then :)

SmallJoker commented on src/content/subgames.cpp:
```diff
@@ -288,8 +306,27 @@ std::vector<WorldSpec> getAvailableWorlds()
+	std::string final_path = path;
+
+	// If we're creating a new world, ensure that the path isn't already taken
+	if (mustCreateNew) {
```
Low priority: According to the guidelines, local variables and arguments should use underscores.

Suggestion:
```suggestion
	if (create_new_world) {
```

SmallJoker commented:
:+1: works still as expected.
